src/snowflake/snowpark/_internal/data_source_utils.py

```diff
@@ -0,0 +1,347 @@
+            raw_schema = cursor.fetchall()
+            return oracledb_to_snowpark_type(raw_schema)
+        else:
+            raise NotImplementedError(
+                f"currently supported drivers are pyodbc and oracledb, got: {current_db}"
+            )
+    except Exception as exc:
+        raise SnowparkDataframeReaderException(
+            f"Unable to infer schema from table {table}"
+        ) from exc
+
+
+def data_source_data_to_pandas_df(
+    data: List[Any], schema: StructType, conn: Connection
+) -> pd.DataFrame:
+    columns = [col.name for col in schema.fields]
+    df = pd.DataFrame.from_records(data, columns=columns)
+
+    # convert timestamp and date to string to work around SNOW-1911989
+    df = df.map(
+        lambda x: x.isoformat()
+        if isinstance(x, (datetime.datetime, datetime.date))
+        else x
+    )
+    # convert binary type to object type to work around SNOW-1912094
+    df = df.map(lambda x: x.hex() if isinstance(x, (bytearray, bytes)) else x)
+    current_db = detect_dbms(conn)
+    if current_db == DBMS_TYPE.SQL_SERVER_DB or current_db == DBMS_TYPE.SQLITE_DB:
+        return df
+    elif current_db == DBMS_TYPE.ORACLE_DB:
+        from oracledb import LOB
+
+        tz_data = []
+        for col in schema.fields:
+            if isinstance(col.datatype, TimestampType) and col.datatype.tz in [
+                TimestampTimeZone.TZ,
+                TimestampTimeZone.LTZ,
+            ]:
+                tz_data.append(col.name)
+        # apply read to LOB object, we currently have FakeOracleLOB because CLOB and BLOB is represented by an
+        # oracledb object and we cannot add it as our dependency in test, so we fake it in this way
+        # TODO: SNOW-1923698 remove FakeOracleLOB after we have test environment
+        df = df.map(lambda x: x.read() if isinstance(x, (LOB, FakeOracleLOB)) else x)
+        for column in tz_data:
+            df[column] = df[column].apply(lambda x: parser.parse(x))
+
+    else:
+        raise NotImplementedError(
+            f"currently supported drivers are pyodbc and oracledb, got: {current_db}"
+        )
+    return df
+
+
+def generate_select_query(table: str, schema: StructType, conn: Connection) -> str:
+    current_db = detect_dbms(conn)
+    if current_db == DBMS_TYPE.ORACLE_DB:
+        cols = []
+        for field in schema.fields:
+            if (
+                isinstance(field.datatype, TimestampType)
+                and field.datatype.tz == TimestampTimeZone.TZ
+            ):
+                cols.append(
+                    f"""TO_CHAR({field.name}, 'YYYY-MM-DD"T"HH24:MI:SS.FF6 TZH:TZM')"""
+                )
+            elif (
+                isinstance(field.datatype, TimestampType)
+                and field.datatype.tz == TimestampTimeZone.LTZ
+            ):
+                cols.append(
+                    f"""TO_CHAR({field.name} AT TIME ZONE SESSIONTIMEZONE, 'YYYY-MM-DD"T"HH24:MI:SS.FF6 TZH:TZM')"""
```

Comment:
`YYYY-MM-DD"T"HH24:MI:SS.FF6 TZH:TZM` to what second precision (.FFx) does oracle db support? I know snowflake supports up to FF9

Reply:
this is also true for oracledb that is support up to FF9, updated